### ISSUE TITLE
Issues with lambdas in new ofEvent style

### ISSUE THREAD

**Issue by mattfelsen** [openframeworks/openFrameworks#5031]
I was playing with the new style of registering for events, following the info in this commit message
https://github.com/openframeworks/openFrameworks/commit/52434595cec8dbd7a6af2930e35c815d4ba77994

Using `ofEvent::newListener()` seems to be working fine for binding callbacks to member functions, both with `ofEvent<void>` and `ofEvent<Object>`, i.e. these work fine:

```
ofEvent<void> testEvent;
ofEvent<string> testStringEvent;

myEventListener = ofApp::testEvent.newListener(this, &ofApp::onTestEvent);
myStringEventListener = ofApp::testEvent.newListener(this, &ofApp::onTestStringEvent);
```

However, trying to pass in a lambda results in a few compile errors. For example, these results in:
```
myEventListener = ofApp::testEvent.newListener(this, [](){});
myStringEventListener = ofApp::testStringEvent.newListener(this, [](string& s){});
```

```
events/ofEvent.h:614:38: No matching member function for call to 'make_function'
events/ofEvent.h:539:14: Candidate template ignored: could not match 'void (TObj::*)(const void *)' against '(lambda at apps/myApps/EventsTest/src/ofApp.cpp:10:58)'
events/ofEvent.h:597:14: Candidate function not viable: requires 2 arguments, but 3 were provided
```

If I try the moseMoved example from the commit message, I see..

```
ofEvents().mouseMoved.newListener([](ofMouseEventArgs & mouse){
    cout << mouse << endl;
});
```

```
events/ofEvent.h:441:38: Call to member function 'make_function' is ambiguous
events/ofEvent.h:363:14: Candidate function not viable: no known conversion from '(lambda at apps/myApps/EventsTest/src/ofApp.cpp:14:39)' to 'bool (*)(ofMouseEventArgs &)' for 1st argument
events/ofEvent.h:338:14: Candidate function template not viable: requires 3 arguments, but 2 were provided
```

**Comment by arturoc** [comment]
yes this is a known problem you should be able to use it if you use a capture by reference:

```cpp
ofEvents().mouseMoved.newListener([&](ofMouseEventArgs & mouse){
    cout << mouse << endl;
});
```

**Comment by mattfelsen** [comment]
Ah, perfect! Thanks for the quick response. And now I realize that in my test events above, I was passing `this` before the lambda, which doesn't make sense. Removing it fixes the issues there. And the last thing worth noting is that even for an `ofEvent<void>` it looks like the lambda still has to capture by reference, i.e. `[](){}` won't work, it needs to be `[&](){}` even though there are no arguments being passed.

I'll put together a PR with an example of the new registering style soon, and maybe include both the "new" and "legacy" styles. @arturoc what are your thoughts on migrating the existing events examples to the new `ofEventListener` style? Is that the new "official" path forward, or should it wait for now?

**Comment by elliotwoods** [comment]
huh? why do you need to pass *everything* as a reference into the lambda function?
can this be fixed?

**Comment by arturoc** [comment]
it's just the default signature the events accept, we can probably just overload it to accept some other formats, probably without reference and without argument at all would be sueful

**Comment by elliotwoods** [comment]
generally i use ` std::function<void(ArgType&)>` as the argument for adding callbacks (for ofxLiquidEvent) which is maybe more permissive with the signature relating to scope (i.e. in my case it allows user to include whatever they like in the scope). 
i had a lot of issues trying to have an a method on the event which supports both `function<void(ArgType&)>` and `function<void()>`. It seems there's some deep ambiguity when trying to overload the method to accept both those data types (e.g. need `addVoidListener` and `addListener` seperately). but perhaps you've got something different going on which fixes that.
